### ISSUE TITLE
Bind to all addresses if host name is unspecified in box.cfg.listen

### ISSUE THREAD

**Issue by locker** [tarantool/tarantool#7152]
When `box.cfg.listen` is set to a port number (without specifying a host name), Tarantool binds to one address, ipv4 or ipv6:

```lua
box.cfg{listen = 3301}
```

```
$ nc -vvz 127.0.0.1 3301
Connection to 127.0.0.1 3301 port [tcp/*] succeeded!
$ nc -vvz ::1 3301
nc: connect to ::1 port 3301 (tcp) failed: Connection refused
```

This is confusing - one would expect it to bind to all available addresses, like most standard commands do, e.g. both `ncat` and `s_server` bind to both ipv4 and ipv6 if the host name is unspecified:

```
ncat -kl 3301
openssl s_server -key server.key -cert server.crt -accept 3301 -www
```

Now, that #5928 is fixed, we could bind to all addresses returned by `getaddrinfo`, not just the first one. Look [here](https://github.com/tarantool/tarantool/blob/ff227f875f5bf7b43671bd02584e9acbdd0c65f1/src/lib/core/evio.c#L356-L373).

**Comment by LevKats** [comment]
It looks like it's not simply binding all `getaddrinfo` results with `evio_service_entry_bind_addr` [here](https://github.com/tarantool/tarantool/blob/ff227f875f5bf7b43671bd02584e9acbdd0c65f1/src/lib/core/evio.c#L365C7-L365C35).
The reason is that we call `evio_service_entry_bind` in `evio_service_bind` [here](https://github.com/tarantool/tarantool/blob/e0c9b6527f7897007028358adef963d542526d80/src/lib/core/evio.c#L533C7-L533C30) after the `evio_service_create_entries` call where we pass `uri_set->uri_count` that results in allocation of that amount of entries [here](https://github.com/tarantool/tarantool/blob/e0c9b6527f7897007028358adef963d542526d80/src/lib/core/evio.c#L446). In case when `evio_service_entry_bind` call may result in binding of several entries we would need to resize that array as each uri may for example bind 2 entries. So we would need to change `evio_service_entry_bind` signature in a way that it returns number of binded entries and constantly realloc `service->entries` taking care of it's real capacity or use some kind of rlist of future entries that would be passed to service->entries after binding. Solutions may require non-trivial change in interfaces.

Also there are some complications with testing of this feature in the future:
1. We should skip the test on systems that are not configured with ipv6
2. We should potentially ignore `TARANTOOL_LISTEN` environmental variable because it may differ from port number
3. We should somehow test it on available port that is not used by some tarantool instance or some other service 
4. What is the expected result of passing  `box.cfg{listen = 0}`? Is it binding free (probably different) ports for each ip? Is it binding the same port for all ips? Or is it binding one port to _some_  ip?

**Comment by Serpentian** [comment]
> It looks like it's not simply binding all getaddrinfo results with evio_service_entry_bind_addr

I agree, this looks way too difficult to implement this in `evio` module and propose to solve the problem on the higher level: e.g. creating several uri entries during parsing the list in `listen`:

https://github.com/tarantool/tarantool/blob/e0c9b6527f7897007028358adef963d542526d80/src/lib/uri/uri.c#L378-L379

---

> We should skip the test on systems that are not configured with ipv6

We may use `t.skip_if` for that:

https://github.com/tarantool/tarantool/blob/e0c9b6527f7897007028358adef963d542526d80/test/app-luatest/gh_5983_jit_library_smoke_tests_test.lua#L20-L21

But we have to find out, using which condition we wanna skip the test. The solution may be introducing a function in test, which will figure out, whether ipv6 is disabled, but it'll have to support Linux and MacOS at least: in linux we can easily check it with `cat /sys/module/ipv6/parameters/disable`. Or we can probably use `socket` module and try binding to ipv6, if it fails, then ipv6 is disabled.

---

> We should potentially ignore TARANTOOL_LISTEN environmental variable because it may differ from port number

In general `TARANTOOL_LISTEN` is a socket file, we cannot use it in our tests. Probably binding to a non-popular port will be ok. Maybe we can even randomize port, so that on restart the test passes, if the port is busy. Note, that on CI flaky tests are restarted. 

---

> What is the expected result of passing box.cfg{listen = 0}? Is it binding free (probably different) ports for each ip? Is it binding the same port for all ips? Or is it binding one port to some ip?

Good question. Port from ipv4 and ipv6 don't intersect (I can easily bind one tnt instance to '127.0.0.1:3301' and another one to '[::1]:3301'). 0 means first available port and it may be different for ipv4 and ipv6. I suppose, these should be different ports, but both of them must be shown in `box.info.listen`. @locker.

---

Another question for @locker. What we're going to do if ipv6 is disabled on computer and user passes just port to `listen`? Now tarantool fails if it cannot bind to at least one of the addresses in `listen`. I'm afraid it will be way too bad to exit, when client specifies port number and we cannot bind to ipv6.

**Comment by locker** [comment]
> So we would need to change `evio_service_entry_bind` signature in a way that it returns number of binded entries and constantly realloc `service->entries` taking care of it's real capacity or use some kind of rlist of future entries that would be passed to service->entries after binding. Solutions may require non-trivial change in interfaces.

Turning `evio_service::entries` into a list is fine. Changing private function signatures is fine, as well. We can refactor the evio service code as we wish.

> I agree, this looks way too difficult to implement this in `evio` module and propose to solve the problem on the higher level: e.g. creating several uri entries during parsing the list in `listen`

I wouldn't do that - this would complicate error handling.

> What is the expected result of passing  `box.cfg{listen = 0}`?

No special handling. Don't attempt to find a free port. Nobody asked for this so far.

> What we're going to do if ipv6 is disabled on computer and user passes just port to `listen`?

Bind only to ipv4, as we do now. That is, don't raise an error if we manager to bind to at least one port.

**Comment by Serpentian** [comment]
> No special handling. Don't attempt to find a free port. Nobody asked for this so far.

It already works like that, 0 - any free port. We should not bind to ipv6 if 0 is set?

**Comment by locker** [comment]
> > No special handling. Don't attempt to find a free port. Nobody asked for this so far.
> 
> It already works like that, 0 - any free port. We should not bind to ipv6 if 0 is set? Seems counterintuitive.

Didn't know that. We don't have any special handling for this - it looks like this is how `getaddrinfo` works. Then we should bind to whatever ports `getaddrinfo` returns (using different ports for ipv4 and ipv6 is fine). In this case `box.info` should report all bind addresses (AFAICS it should work as expected without any additional patching because it retrieves the info from `evio_service`).

**Comment by Totktonada** [comment]
It seems, solving this problem would be useful for #10560.
